src/client.js

```diff
@@ -0,0 +1,275 @@
+import assert from 'assert';
+import crypto from 'crypto';
+
+import fs from 'mz/fs';
+
+import Runner from './runner';
+import InterchangeFormat from './interchange-format';
+import {Joi, schemas, runSchema, MB, GB, TB} from './schemas';
+
+const MAX_S3_CHUNKS = 10000;
+
+/**
+ * This class represents a client of a service implemented using the
+ * `Controller` class.
+ */
+class Client {
+
+  constructor(opts) {
+    opts = runSchema(opts || {}, Joi.object().keys({
+      runner: Joi.any(),
+      runnerOpts: Joi.object().default({}),
+      partsize: Joi.number().min(5*MB).max(5*GB).default(25*MB),
+      multisize: Joi.number().default(100*MB),
+    }).without('runner', 'runnerOpts')
+      .optionalKeys(['runner', 'runnerOpts']));
+
+    let {runner, runnerOpts, partsize, multisize} = opts;
+
+    // Store value and not reference
+    this.partsize = partsize+0;
+
+    // Unlike the Controller, which has much simplier usage, here we do want
+    // the full Runner api to be available to the client
+    if (!runner) {
+      runner = new Runner(opts.runnerOpts);
+    }
+    
+    // The Runner to use for this
+    this.runner = runner;
+
+    // The minimum size before switching to multipart
+    this.multisize = opts.multisize;
+  }
+
+
+  async __prepareSinglepartUpload(opts) {
```

Comment:
Well, technically speaking a single part upload is either a PUT to the object *or* a multipart upload with a single part.  Since those are two distinct things under the hood, I'd rather represent them as two different things in the client.  In the final command line tools and libraries that are TC specific, this will all be hidden away.